dispatcher module for multiple graphs

avikchaudhuri opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #139439
* #139438

Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

This PR introduces the concept of a "dispatcher" module `n` that carries multiple interpreter modules `n`, `n@1`, `n@2`, etc., each corresponding to a particular call of `n` and thus might carry a different specialized graph. We only do this when we're preserving module call signatures for `n`. The carried modules have the same number and order of calls to `n` appearing in the original module / exported program. In the unflattened module, all those calls go to the "dispatcher" module which internally tracks how many calls have been made so far and invokes the corresponding interpreter module. We reset this tracking after a successful or unsuccessful run of the unflattened module.

Overall this makes swapping easier when module call signatures are preserved.

avikchaudhuri committed:
dispatcher module for multiple graphs



Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "dispatcher module for multiple graphs"


Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "dispatcher module for multiple graphs"


Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

[ghstack-poisoned]

avikchaudhuri committed:
Update on "dispatcher module for multiple graphs"


Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

[ghstack-poisoned]

tugsbayasgalan commented on test/export/test_export.py:
```diff
@@ -6620,6 +6620,44 @@ def forward(self, x):
+        eager_result = m(*inp)
+
+        if not is_retracebility_test(self._testMethodName):
+            ep = export(M(), inp, preserve_module_call_signature=("n",))
```
Can you test it with just training IR as well?

avikchaudhuri replied:
OK.

avikchaudhuri committed:
Update on "dispatcher module for multiple graphs"


Differential Revision: [D65307961](https://our.internmc.facebook.com/intern/diff/D65307961/)

This PR introduces the concept of a "dispatcher" module `n` that carries multiple interpreter modules `n`, `n1`, `n2`, etc., each corresponding to a particular call of `n` and thus might carry a different specialized graph. We only do this when we're preserving module call signatures for `n`. The carried modules have the same number and order of calls to `n` appearing in the original module / exported program. In the unflattened module, all those calls go to the "dispatcher" module which internally tracks how many calls have been made so far and invokes the corresponding interpreter module. We reset this tracking after a successful or unsuccessful run of the unflattened module.

Overall this makes swapping easier when module call signatures are preserved.

[ghstack-poisoned]
